server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -909,132 +916,139 @@ public void snapshotShard(Store store, MapperService mapperService, SnapshotId s
+                    }
+                    // delete all files that are not referenced by any commit point
+                    // build a new BlobStoreIndexShardSnapshot, that includes this one and all the saved ones
+                    List<SnapshotFiles> newSnapshotsList = new ArrayList<>();
+                    newSnapshotsList.add(new SnapshotFiles(snapshot.snapshot(), snapshot.indexFiles()));
+                    for (SnapshotFiles point : snapshots) {
+                        newSnapshotsList.add(point);
+                    }
+                    final String indexGeneration = Long.toString(fileListGeneration + 1);
+                    final List<String> blobsToDelete;
+                    try {
+                        final BlobStoreIndexShardSnapshots updatedSnapshots = new BlobStoreIndexShardSnapshots(newSnapshotsList);
+                        indexShardSnapshotsFormat.writeAtomic(updatedSnapshots, shardContainer, indexGeneration);
+                        // Delete all previous index-N blobs
+                        blobsToDelete =
+                            blobs.keySet().stream().filter(blob -> blob.startsWith(SNAPSHOT_INDEX_PREFIX)).collect(Collectors.toList());
+                        assert blobsToDelete.stream().mapToLong(b -> Long.parseLong(b.replaceFirst(SNAPSHOT_INDEX_PREFIX, "")))
+                            .max().orElse(-1L) < Long.parseLong(indexGeneration)
+                            : "Tried to delete an index-N blob newer than the current generation [" + indexGeneration
+                            + "] when deleting index-N blobs " + blobsToDelete;
+                    } catch (IOException e) {
+                        throw new IndexShardSnapshotFailedException(shardId,
+                            "Failed to finalize snapshot creation [" + snapshotId + "] with shard index ["
+                                + indexShardSnapshotsFormat.blobName(indexGeneration) + "]", e);
+                    }
+                    try {
+                        shardContainer.deleteBlobsIgnoringIfNotExists(blobsToDelete);
+                    } catch (IOException e) {
+                        logger.warn(() -> new ParameterizedMessage("[{}][{}] failed to delete old index-N blobs during finalization",
+                            snapshotId, shardId), e);
+                    }
+                    context.finish(threadPool.absoluteTimeInMillis());
+                }, onFailure), indexIncrementalFileCount + 1);
+            filesListener.onResponse(null);
+            final Executor executor = threadPool.executor(ThreadPool.Names.SNAPSHOT);
+            for (BlobStoreIndexShardSnapshot.FileInfo snapshotFileInfo : filesToSnapshot) {
+                executor.execute(new ActionRunnable<>(filesListener) {
+                    @Override
+                    protected void doRun() {
+                        try {
+                            snapshotFile(snapshotFileInfo, indexId, shardId, snapshotId, snapshotStatus, store);
```

Comment:
@tlrx yea that sounds like a reasonable approach for sure. The only issue with that, that I can see is this snippet in `snapshotFile`:

```
        } catch (Exception t) {
            failStoreIfCorrupted(store, t);
            snapshotStatus.addProcessedFile(0);
            throw t;
```

-> it is a little tricky to convert this into what you propose because we need the error but also the effect of a `0` return. Maybe we should make this nicer in a follow up? (also keeps the change complexity down a little to not change the abort mechanics here?)

Reply:
The usage of snapshotStatus.addProcessedFile(0); should be the responsability of the caller, which should catch any exception.

+1 for a follow up